### PR TITLE
feat!: implement expired subspace fee grant deletion

### PR DESCRIPTION
**Pull request by dadamu**
## Description

This PR implements expired subspace fee grant deletion mentioned in [ADR-017](https://github.com/desmos-labs/desmos/blob/master/docs/architecture/adr-017-subspace-fee-grant.md#store-1).

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/desmos-labs/desmos/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://docs.cosmos.network/v0.44/building-modules/intro.html)
- [ ] included the necessary unit and integration [tests](https://github.com/desmos-labs/desmos/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/subspaces/abci.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/desmos-labs/desmos/v5/x/subspaces/keeper"
+)
+
+// BeginBlocker called every block, remove expired allowances
+func BeginBlocker(ctx sdk.Context, keeper keeper.Keeper) {
+	// Remove allowances that have been expired by the current block time
+	keeper.RemoveExpiredAllowances(ctx, ctx.BlockTime())
+}
```
Comment (RiccardoM) [inline x/subspaces/abci.go:13]:
```suggestion
// BeginBlocker is called every block and takes care of removing expired allowances
func BeginBlocker(ctx sdk.Context, keeper keeper.Keeper) {
	keeper.RemoveExpiredAllowances(ctx, ctx.BlockTime())
}
```

File: x/subspaces/keeper/feegrant.go [hunk 2]
```diff
@@ -87,6 +102,50 @@ func (k Keeper) GetGroupGrant(ctx sdk.Context, subspaceID uint64, groupID uint32
 
 // --------------------------------------------------------------------------------------------------------------------
 
+// saveAllowanceToExpiringQueue saves the allowance into expiring queue
+func (k Keeper) saveAllowanceToExpiringQueue(ctx sdk.Context, expiration *time.Time, grantKey []byte) {
+	k.removeAllowanceFromExpiringQueue(ctx, grantKey)
```
Comment (RiccardoM) [inline x/subspaces/keeper/feegrant.go:107]: Why is this needed?

File: x/subspaces/keeper/feegrant.go [hunk 3]
```diff
@@ -41,6 +48,10 @@ func (k Keeper) HasUserGrant(ctx sdk.Context, subspaceID uint64, grantee string)
 func (k Keeper) DeleteUserGrant(ctx sdk.Context, subspaceID uint64, grantee string) {
 	store := ctx.KVStore(k.storeKey)
 	key := types.UserAllowanceKey(subspaceID, grantee)
+
+	k.removeAllowanceFromExpiringQueue(ctx, key)
```
Comment (RiccardoM) [inline x/subspaces/keeper/feegrant.go:52]:
```suggestion
	// Remove the allowance from the list of expiring allowances
	k.removeAllowanceFromExpiringQueue(ctx, key)
```

File: x/subspaces/keeper/feegrant.go [hunk 4]
```diff
@@ -68,6 +79,10 @@ func (k Keeper) HasGroupGrant(ctx sdk.Context, subspaceID uint64, groupID uint32
 func (k Keeper) DeleteGroupGrant(ctx sdk.Context, subspaceID uint64, groupID uint32) {
 	store := ctx.KVStore(k.storeKey)
 	key := types.GroupAllowanceKey(subspaceID, groupID)
+
+	k.removeAllowanceFromExpiringQueue(ctx, key)
```
Comment (RiccardoM) [inline x/subspaces/keeper/feegrant.go:83]:
```suggestion
	// Remove the allowance from the list of expiring allowances
	k.removeAllowanceFromExpiringQueue(ctx, key)
```

File: x/subspaces/keeper/feegrant_test.go [hunk 5]
```diff
@@ -140,6 +142,200 @@ func (suite *KeeperTestSuite) TestKeeper_SaveGrant() {
...
+			),
+			check: func(ctx sdk.Context) {
+				suite.Require().True(suite.k.HasGroupGrant(ctx, 1, 1))
+
+				// Check no grants inside expiration queue
+				suite.Require().Empty(suite.getAllGrantsInExpiringQueue(ctx))
+			},
+		},
+		{
+			name: "new grant with expiration is saved properly properly",
```
Comment (RiccardoM) [inline x/subspaces/keeper/feegrant_test.go:173]:
```suggestion
			name: "new grant with expiration is saved properly",
```

File: x/subspaces/types/models_feegrant_test.go [hunk 6]
```diff
@@ -165,3 +166,50 @@ func TestGrant_Validate(t *testing.T) {
 		})
 	}
 }
+
+func TestGrant_GetExpiration(t *testing.T) {
```
Comment (RiccardoM) [inline x/subspaces/types/models_feegrant_test.go:170]: Can we add a test case with an allowance that does *not* have an expiration to make sure the method does not panic?

File: x/subspaces/legacy/v7/store.go [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var grant types.Grant
+		cdc.MustUnmarshal(iterator.Value(), &grant)
+
+		expiration := grant.GetExpiration()
+		if expiration != nil {
+			allowanceKey := allowancePrefix
+			allowanceKey = append(allowanceKey, iterator.Key()...)
```
Comment (RiccardoM) [inline x/subspaces/legacy/v7/store.go:34]: These lines can be combined
```suggestion
			allowanceKey := append(allowancePrefix, iterator.Key()...)
```

Comment (dadamu) [reply x/subspaces/legacy/v7/store.go:34]: It would cause the linter error: `appendAssign: append result not assigned to the same slice (gocritic)`.
https://github.com/desmos-labs/desmos/actions/runs/5552100585/jobs/10139013465#step:5:146

File: x/subspaces/keeper/feegrant.go (hunk 2 above)
Comment (dadamu) [reply x/subspaces/keeper/feegrant.go:109]: If a grant updated, we should make sure the grant with original expiration is removed first. 

For instance, there is a grant expired at `20220707` and we would like to update the expiration time to `20220708`, the grant will have two expiring key if it is not removed the old one first, like:
```
expiring key | 20220707 | grant key
expiring key | 20220708 | grant key
```

File: x/subspaces/legacy/v7/store.go (hunk 7 above)
Comment (RiccardoM) [reply x/subspaces/legacy/v7/store.go:34]: What about something like this?
```go
allowanceKey := append([]byte(allowanceKey...), iterator.Key()...)
```

File: x/subspaces/types/models_feegrant_test.go (hunk 6 above)
Comment (dadamu) [reply x/subspaces/types/models_feegrant_test.go:170]: Added test

File: x/subspaces/keeper/feegrant.go (hunk 2 above)
Comment (RiccardoM) [reply x/subspaces/keeper/feegrant.go:109]: Ok I see. I think it's better to add a comment then. Something like: 

```
// Make sure we remove the grant from the expiring queue to properly handle expiration date updates and avoid duplicated keys
```

File: x/subspaces/types/models_feegrant_test.go (hunk 6 above)
Comment (RiccardoM) [reply x/subspaces/types/models_feegrant_test.go:179]: I meant having a test for a grant with no expiration time. That means a grant that will never expire (`ExpiresAt` that returns `nil`). I don't think we need to test the case in which `ExpiresAt` returns an error hoestly

File: x/subspaces/legacy/v7/store.go (hunk 7 above)
Comment (dadamu) [reply x/subspaces/legacy/v7/store.go:34]: It would be another linter problem: 
``
x/subspaces/legacy/v7/store.go:33:33: unnecessary conversion (unconvert)
                        allowanceKey := append([]byte(allowancePrefix), iterator.Key()...)
``

File: x/subspaces/types/models_feegrant_test.go (hunk 6 above)
Comment (dadamu) [reply x/subspaces/types/models_feegrant_test.go:170]: Ah, I see, apologize the misunderstanding.

Added the case which is `no expiration returns nil`.

File: x/subspaces/keeper/feegrant.go (hunk 2 above)
Comment (dadamu) [reply x/subspaces/keeper/feegrant.go:110]: Added the comment

File: x/subspaces/legacy/v7/store.go (hunk 7 above)
Comment (RiccardoM) [reply x/subspaces/legacy/v7/store.go:34]: Ok then we can leave it as it is

Comment (dadamu) [reply x/subspaces/legacy/v7/store.go:34]: I refer to the similar codes which uses `append([]byte(nil), prefix)` pattern. I think we can apply it. 

https://github.com/desmos-labs/desmos/blob/master/x/profiles/keeper/grpc_query.go#L179

**Review by RiccardoM — APPROVED** [review]
